### ISSUE TITLE
rar2fs process failing on Windows 10/Cygwin 64

### ISSUE THREAD

**Issue by janosek** [hasse69/rar2fs#81]
Hello,

I use rar2fs in Cygwin 64 to mount a folder full of RAR files in Windows 10.  The program executes and all is good.  When I come back a few hours later, the program is gone.  I execute the command to load and mount the file system again, and it is back.  I created a log file and I got one interesting item:
      0 [main] rar2fs 20196 C:\cygwin64\usr\local\bin\rar2fs.exe: *** fatal erro
r in forked process - WFSO timed out after longjmp
 153271 [main] rar2fs 20196 cygwin_exception::open_stackdumpfile: Dumping stack trace to rar2fs.exe.stackdump

Which contains...

Stack trace:
Frame        Function    Args
000FFFFC6C0  0018005CC8C (000FFFFE3F4, 00000000008, 00100000004, 000FFFFDE50)

I would be happy to provide more diagnostic information, but I may need to be instructed how.

**Comment by hasse69** [comment]
Thanks for the issue report.
What variant of the fuse library are you using in your Cygwin setup? WinFSP? Dokany?
Can you confirm that this crash is seen after simply leaving the mount point idle without any actions towards it?

**Comment by janosek** [comment]
Hello,

I am using WinFSP.  I compiled everything following your back and forth with billziss-gh in the Issue 75 thread.

I also find it fails silently as well.  I just got back to my computer and the process was gone.  The log from the bash script showed nothing.  To be sure, I have Serviio attacking the mount point.  It works well when it is running.  I just want it to be stable.

**Comment by hasse69** [comment]
Right. This is going to be a tricky one to attack. First we need to narrow it down a bit, but there are a few things we can do to accomplish that at least.

First of all are you mounting to a Windows drive letter or to a local folder?

Can you try to run rar2fs in the foreground using the `-d` switch? That would possibly output something of use to use at the time of death.

Another thing you can do is to monitor the system regularly *before* it takes the decision to break apart.
Looking at the output from `ps` or `top` could reveal something, at least if it has to do with spawning of new processes etc. If this is about resource shortage it could tell us a lot. Try to see if the process list is growing for some reason without ever attempting to drop to a sane level. I do not really know what `Serviio` is so an explanation of how it is accessing the mount point would help.

I assume that if the mount point is empty or you do not run `Serviio` you will not see this problem?
Limiting the mount point to hold only a few archives or only archives in store mode (non-compressed) could also provide useful information. Compressed archives requires more resources.

**Comment by janosek** [comment]
Hello,

I execute the shell script as follows in a windows task schedule on start up.
c:\cygwin64\bin\bash.exe -l -c "/home/janosek/rar2fs_start.sh >> /home/janosek/rar2fs.log 2>&1"


rar2fs_start.sh:
(I just added the -d option in the shell script below.)
#!/bin/bash
rar2fs  -d -f /cygdrive/d/files/seed  -o uid=-1,gid=-1 -o 'VolumePrefix=\\Janosek-pc\\x' X:

Serviio is a media server, like Plex.  
I set it to monitor \\Janosek-pc\X for new videos to process.  It works very well when it is active.

**Comment by hasse69** [comment]
I was lazy and did not checkout `Serviio`, sorry for that. But I sort of guessed it was a media server of some sort. So my best guess it is some periodic scanning/indexing that fails. Please try what I asked about before and most importantly try to remove any archives that are compressed. If `Cygwin` does not like the way `rar2fs` is forking we need to look into it of course. However, I do not think `rar2fs` is in error here really, rather `Cygwin` is not handling the fork properly or runs out of resources.

**Comment by hasse69** [comment]
Please try this patch on master/HEAD. It might make a huge difference so please try it out and let me know,

[fork.patch.txt](https://github.com/hasse69/rar2fs/files/1235916/fork.patch.txt)

**Comment by janosek** [comment]
I am out of town at the moment.  I will definitely try this early next week.. Thanks!

**Comment by janosek** [comment]
OK.. I grabbed level 1 from GIT (googled how to grab the head in GIT), everything compiled but I have an unrelated environment issue with Cygwin 64.  It started when I was playing around with some debug tools.

Software Install@Janosek-PC /usr/local/bin
$ ls -ltr
total 820
-rwxrwxrwx 1 Software Install None 750778 Aug 21 07:39 rar2fs.exe
-rwxrwxrwx 1 Software Install None  82549 Aug 21 07:39 mkr2i.exe

Software Install@Janosek-PC /usr/local/bin
$ ./rar2fs.exe
-bash: ./rar2fs.exe: No such file or directory

Software Install@Janosek-PC /usr/local/bin
$ rar2fs.exe
-bash: /usr/local/bin/rar2fs.exe: No such file or directory


Why  do I get no such file or directory when the file is clearly there and clearly executable?

**Comment by hasse69** [comment]
Yea, this is a very unfortunate issue with Cygwin it seems. I have no clue why it happens and what is causing it. I experienced something similar with the `make.exe` tool a while back. The only thing that could work around it was a clean and fresh install of Cygwin. I suggest you post to the Cygwin support forum. It would be very interesting to have someone explain what is going on here.

**Comment by janosek** [comment]
Still have the problem.  I deleted the cygwin64 folder.  Completely removed all registry references to cygwin64 and then installed a fresh copy in c:\cygwin.  Still the same problem.  I will contact the support forum.

**Comment by janosek** [comment]
FIXED IT!

Believe it or not, the problem was a registry entry for windbg.  I was messing with it to try and figure out if I could get a core dump when the process died.  I deleted the entry, re-ran cygwin and typed rar2fs and it worked right away!

I will get my environment set up and test the latest built.
Just to be clear, I did a git clone --depth=1, so hopefully I got that change.

**Comment by hasse69** [comment]
You will not get the change. You need to apply the patch I attached over your cloned master/HEAD.

**Comment by janosek** [comment]
Ahh.. OK I just ran "git apply fork.patch.txt"

**Comment by janosek** [comment]
At the risk of speaking too soon, things look a lot more stable now.  I woke up to rar2fs still being active.  This morning, I reset the serviio folder to the rar2fs mount, so it wiped the existing library and is now re-indexing it, so this should be a suitable stress test.  I also noticed in procexp64 that the threads are now stable at around 14, whereas before it would climb to 21 with threads being created and destroyed frequently.

**Comment by janosek** [comment]
OK.. the process was dead when I got home. The log file showed:
rar2fs: PID 12664: mounted X:
The service rar2fs has been started.
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files
scandir: Too many open files

**Comment by hasse69** [comment]
> scandir: Too many open files
[REPEATED MANY TIMES]

I think in this case you simply hit a rather low default limit in cygwin for maximum number of open files.
Looking at my own cygwin installation `ulimit` give me this:
```
$ ulimit -a
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
open files                      (-n) 256
pipe size            (512 bytes, -p) 8
stack size              (kbytes, -s) 2036
cpu time               (seconds, -t) unlimited
max user processes              (-u) 256
virtual memory          (kbytes, -v) unlimited
```

You can try to increase this before starting `rar2fs` using:
```
$ ulimit -n 3200
$ ulimit -n
3200
```

Note that you cannot increase any higher than this. It is a built-in limitation in `Cygwin`.
The default limit of 256 is actually rather low, especially for an application doing indexing of an entire library, like `serviio` is doing in this case. Also note that you need to keep setting this each time you run a new `Cygwin` shell. I am not sure how to (if even possible) you can make this the default at startup.

**Comment by billziss-gh** [comment]
There is nothing in WinFsp that would terminate a process (other than a possible crashing bug of course, but I am not aware of any).

Looking at the rest of this issue I see that the original issue has to do with forking. Does rar2fs fork a lot (after the file system starts)? Cygwin's fork is rather fragile and the presence of the non-Cygwin WinFsp DLL complicates matters.

**Comment by hasse69** [comment]
@billziss-gh 
> Does rar2fs fork a lot (after the file system starts)?
Well, that is essentially what the patch I provided in this thread is addressing. It used to fork a lot but now use a strict threading model instead. I think the original issue was caused by the possibly limited fork support in Cygwin and we are now hitting a new problem. Still cannot understand though why breaching the number of open files limit would terminate the process. Especially since it seems to handle the breach numerous times before giving up.

**Comment by janosek** [comment]
To respond to the ulimit.  I added ulimit -n 3200 to my shell script:

rar2fs_start.sh
*****************************
#!/bin/bash 

ulimit -n 3200 
rar2fs  -d -f /cygdrive/d/files/seed -o uid=-1,gid=-1 -o 'VolumePrefix=\\Janosek-pc\\x' X: 
****************************

Then from Windows, I execute the following:
c:\cygwin\bin\bash.exe -l -c "/home/janosek/rar2fs_start.sh >> /home/janosek/rar2fs.log 2>&1"

That's how I can set the ulimit each time I execute rar2fs.


I tried to add -o DebugLog=/home/janosek/fuse.log but when I did that, serviio was no longer able to see the volume (even though it otherwise worked fine).  Is there any reason why adding another -o command to the command line would cause the process to function differently? I was hoping I could give you more detailed logs right before the crash, but the program seems to behave differently when I add that string.

**Comment by janosek** [comment]
Still the same problem.

3200
rar2fs: PID 15120: mounted X:
The service rar2fs has been started.
scandir: Too many open files
[repeated many times]
then the process disappears from my Windows taskbar.

**Comment by janosek** [comment]
I am monitoring  /proc/<PID>/fd folder.  The number of open files is increasing.  Started in the 20s, Now over 200.

BTW, is this relevant?:
https://stackoverflow.com/questions/870173/is-there-a-limit-on-number-of-open-files-in-windows

**Comment by hasse69** [comment]
@janosek 
> Is there any reason why adding another -o command to the command line would cause the process to function differently?

Maybe. I am not sure, but AFAIK the `-o` option can not be repeated, one will override the other. You should combine them instead, like `-o uid=-1,gid=-1, 'VolumePrefix=\Janosek-pc\x, DebugLog=/home/janosek/fuse.log`. But I might be wrong of course.

> BTW, is this relevant?:
https://stackoverflow.com/questions/870173/is-there-a-limit-on-number-of-open-files-in-windows

Not sure. But please keep monitoring `/proc/<pid>/fd`. To me it looks like changing the limit of number of open files through `ulimit` has not effect?

**Comment by janosek** [comment]
I am at 2733 files now in the FD folder (ls -1 | wc -l).  They are pointing to tons and tons of RAR files.

**Comment by hasse69** [comment]
@janosek 
> I am at 2733 files now in the FD folder (ls -1 | wc -l). They are pointing to tons and tons of RAR files.

Just to make sure. By this you mean the source .rar and possible volume files like .rNN and .partN.rar?
If that is the case I can hardly blame `serviio` for keeping files open. If it is RAR files then it actually points to `rar2fs`, possibly indirect through `libunrar`.

**Comment by janosek** [comment]
Yes,  all the part RAR files for each set of rars.

**Comment by hasse69** [comment]
@janosek 
> Yes, all the part RAR files for each set of rars.

Thanks. Now the interesting thing here is why I do not see this under Linux? At least I assume would get numerous reports if this was the case.

**Comment by hasse69** [comment]
Can you tell if the open files are only related to RAR archives in compressed format or both compressed and non-compressed?

**Comment by janosek** [comment]
Virtually all files (cannot guarantee 100% though) are uncompressed.

**Comment by janosek** [comment]
I compiled with unrarsrc-5.5.7.tar.gz, if that matters.

**Comment by hasse69** [comment]
Ok, I have found one issue related to the patch. Files are not being closed properly unless the complete file is read. Which confuses me a bit since you are saying you see only files related to archives in store format (uncompressed) but the patch only affects compressed archives.

**Comment by janosek** [comment]
That looks a lot better!

If I keep doing an "ls -ltr" in the /proc/\<PID>/fd folder,  I see new files appearing but then disappearing the next time I run the command

**Comment by janosek** [comment]
I will check in in a few days if it continues to be stable..

So far, the threads look stable at around 16 in Windows and the number of open files is 3 or 4 max.  Fingers crossed this is resolved!

**Comment by janosek** [comment]
Well.. spoke too soon.  
As soon as Serviio got to the spot it left off and started indexing again, the files started to increase.  Let me know what information you need now.

**Comment by janosek** [comment]
From what I can see it leaves open either the last file in the set of RARs, or non-rar files (mp4), open.

**Comment by hasse69** [comment]
@janosek 
>That looks a lot better!

Well it should not really work well because this is the version that is using `fork()` heavily.
But most likely it has nothing to do with forking, but is more related to the number of open files.

>From what I can see it leaves either the last file in the set of RARs, or non-rar files (mp4), open.

This is actually very interesting information. There is only one thing in rar2fs that would open a non-rar file like this. And that is the pass-through part of the file system which only operates on files outside of an archive. Also the fact that it always seems to be the last rar file being left open also indicates that rar2fs does not get the close/release calls it expects. For compressed files `libunrar` handles the release of files once extraction completes. That is not depending on the user closing what ever it is accessing. But for archives in store mode it is handled internally in rar2fs and closing of last file in a volume is always performed as a result of the user releasing whatever file it is accessing inside the archive. Basically saying either `serviio` does not behave nice (leave files open) or there is something in the fuse layer preventing rar2fs from getting the release calls it expects. If this theory is right the files you should see getting stuck in limbo is files not part of an archive and/or files in uncompressed archives. The reason why I expect you see the last rar file in a volume is because that is basically what indexing does, It tries to read the index table at the end of each file.

**Comment by hasse69** [comment]
Ok, there is something really weird going on here in Cygwin! I added a few very dumb printouts from the open() and release() calls in rar2fs and this is what I get for a simple `find` executing `ls` for files ending with `*sample.avi`.

ZZZ OPEN /some/path1/some-sample.avi --> 22
ZZZ OPEN /some/path1/some-sample.avi --> 23
ZZZ RELEASE --> 23
ZZZ OPEN /some/path2/some-sample.avi --> 24
ZZZ OPEN /some/path2/some-sample.avi --> 25
ZZZ RELEASE --> 25
...

As you can see file is opened *twice* but only closed once! Each one of these results in an obvious leak. Exactly the same command on Linux does not result in any files being opened at all. If I instead execute `tail` on them Linux opens the file once and then release it properly, same `fd` is being allocated every time and there are no leaks. I am not sure what to do with this one to be honest, maybe @billziss-gh has some idea of why Cygwin behaves like this or if there is something odd going on in the fuse layer causing open to be called twice but with only only one release/flush.

@billziss-gh 
I see the below pattern all the time in Cygwin. It starts of ok, but then suddenly starts to open files more than once, and only one flush/release call :(
`
ZZZ OPEN /some/path1/sample.avi --> 5
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
ZZZ OPEN /some/path2/sample.avi --> 5
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
ZZZ OPEN /some/path3/sample.avi --> 5
ZZZ OPEN /some/path3/sample.avi --> 6
ZZZ FLUSH --> 6
ZZZ RELEASE --> 6
ZZZ OPEN /some/path4/sample.avi --> 6
ZZZ OPEN /some/path4/sample.avi --> 7
`

**Comment by hasse69** [comment]
I can get this problem simply by doing a recursive `ls` on a mount point with plenty of files. 
It starts of good but then suddenly some release calls are not coming any more and from that point it starts to leak one file descriptor for every file. Doing a simple `ls` on a file give this pattern:

```
ZZZ OPEN --> 3
ZZZ FLUSH --> 3
ZZZ RELEASE --> 3
ZZZ OPEN --> 3
ZZZ FLUSH --> 3
ZZZ RELEASE --> 3
``` 

Which is ok, I do not really care about the fact that files are opened simply because they are listed. Seems strange but other than that, no. Doing a simple `ls` repeatedly does not seem to trigger the problem. But a recursive `ls -R` does. After a while it starts to open files before flushing the previous one, which is fine as long as they are all flushed/released. Which also seems to be the case for a while at least.

```
ZZZ OPEN --> 4
ZZZ OPEN --> 5
ZZZ FLUSH --> 4
ZZZ RELEASE --> 4
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
```

And then eventually it starts to fall apart...

```
ZZZ OPEN  --> 4
ZZZ OPEN  --> 5
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
ZZZ OPEN  --> 5
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
ZZZ OPEN  --> 5
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
ZZZ OPEN  --> 5
ZZZ OPEN  --> 6
ZZZ FLUSH --> 5
ZZZ RELEASE --> 5
ZZZ FLUSH --> 6
ZZZ RELEASE --> 6
```

As you can see, out of nowhere fd 4 is suddenly completely forgotten about. This then starts to happen also for fd 5 and so on. In this run I ended up with 46 file descriptors stuck in limbo.

**Comment by janosek** [comment]
Well the good news, from my end anyway, is my problem is reproducible.  Let me know if you need me to do anything more at this point.  I wish you luck resolving this, because having rar2fs work in Windows is really cool and I imagine more users will attempt similar usage to me in the coming months.

**Comment by janosek** [comment]
> How can this ever be a good thing? :(
(you have a completely useless file system right now)
> 

Its a good thing because if you can reproduce it, you can trace it, even if it is hard (and lets face it, most programmers love solving hard problems).  If only I experienced it and I could not convey what was going on, you would think I am a dunce who wasted your time and I would still have a completely useless file system.  Also, last year this time I looked at rar2fs and longed to have a solution like that for Windows.  Now I have a mostly working solution with some bugs.  This is excellent progress.

> Perhaps. But my headache right now is not how to solve this but rather to understand where the issue resides, because currently I cannot see this being a problem that can be addressed in rar2fs. As long as release() is not called as expected there are going to be resource leaks.
> 
Well, this sound like a serious enough problem that if you can isolate the source, the developer of the source where the problem resides will likely jump on this pretty quickly.

**Comment by hasse69** [comment]
> Its a good thing because if you can reproduce it, you can trace it, even if it is hard (and lets face it, most programmers love solving hard problems)

Sure. I was more thinking from your perspective as a user. As a developer nothing can be better than a bug that can be reproduced. Maybe no bugs at all, but how boring would that be :)

I actually believe I can create a really dirty short term workaround for this issue. It is going to reduce the performance of the file system a lot, but hopefully it will at least not just terminate as it does now.

**Comment by tfiskgul** [comment]
If you are going to reduce the performance by making extra checks or
keeping track of open resources, please consider making it a command line
switch or mount option. So that if one does suffer from this issue, one can
have the choice of enabling "workaround mode".

**Comment by hasse69** [comment]
@tfiskgul 
> If you are going to reduce the performance by making extra checks or
keeping track of open resources, please consider making it a command line
switch or mount option

Of course. But remember, this is an issue affecting Cygwin users only and I would be very surprised if not everyone of those are affected. If, and only if,  there is a workaround, it will apply to Cygwin builds only.

**Comment by hasse69** [comment]
I my attempts to figure out what is going on I have managed to find a rather serious issue with the Cygwin implementation of `pread(2)`. At EOF pread() should return 0, but it does not! It returns some undefined or bogus value. In my case `1628127168`. This seems to confuse `dokany-fuse` to the extent that it enters an endless death-roll loop. A workaround for now is to check if the result from pread is greater than than the requested size and in that case return 0. That prevents the death-roll at least in `dokany-fuse`. I still have to check what happens using `winfsp-fuse`. The reason why I am mentioning `dokan-fuse` here in the first place is because I wish to compare the behavior compared to WinFSP using the same test case. And it is indeed different. There seems to be no file descriptor leaks what so ever using `dokan-fuse`.

@billziss-gh Could it be that there in fact is an issue within the fuse logic in WinFSP? Not saying there is but the fact that behavior is different could need some further investigation? Or what do you think?
I will also check what happens after applying the pread workaround also in the WinFSP case.

**Comment by billziss-gh** [comment]
@hasse69 my apologies, I am still on vacation and cannot properly research this issue.

I am not aware of any file descriptor leak, but this does not mean that there are not any. Be mindful though that Windows (i.e. the OS) will in many cases keep files open for a very long time. This is especially true in WinFsp which supports memory mapping and the cache manager. There is a program called [RAMMap](https://docs.microsoft.com/en-us/sysinternals/downloads/rammap), which allows you to empty the "standby" list. Please try the program and see if it improves this issue (i.e. you suddenly get all the missing RELEASE calls).

In the past (prior to WinFsp v1 release I think) I had a file descriptor leak that manifested itself as follows: The file system would properly open a file and return its file descriptor. For some reason (e.g. sharing violation) the driver would subsequently fail the OPEN request; the driver should then send a RELEASE request to the file system, but it never did. It might be that we have a similar situation here, although I would have thought that by now all such failure scenarios have been resolved.

**Comment by hasse69** [comment]
@billziss-gh 

Thanks for taking your time during vacation!

> There is a program called RAMMap, which allows you to empty the "standby" list. Please try the program and see if it improves this issue (i.e. you suddenly get all the missing RELEASE calls).

Sure, I can see if I can tryout RAMMap. But would this then not be the same issue also for dokan-fuse? 
Using dokan-fuse I seem to get all the expected release calls?
I will also try to reproduce this issue using a simple sample like fusexmp. Would be easier for you to use when digging deeper into the problem I think.

**Comment by billziss-gh** [comment]
> But would this then not be the same issue also for dokan-fuse?
Using dokan-fuse I seem to get all the expected release calls?

Perhaps although Dokan and WinFsp are quite different.

> Btw, should I file a separate issue for this on WinFSP?

Please try RAMMap and feel free to open a WinFsp issue if it does not fix the problem.

**Comment by hasse69** [comment]
@billziss-gh 

> Perhaps although Dokan and WinFsp are quite different.

Well yes, Dokan and WinFSP are different, but the fuse layer should/must not be . They might be complete different implementations (which I am sure they are) but the semantics must be the same to follow the fuse API.

> Please try RAMMap and feel free to open a WinFsp issue if it does not fix the problem.

I will. But it will only give hints to what might be wrong. Please note that file descriptor leak within Cygwin is only one manifestation of the problem with missing release calls. There are plenty of other resources allocated by my files system at open() that must be release properly. A lazy close must never delay release calls to the extent that resources are drained. Default in Cygwin is 256 open files per process. As you can see in this thread it was increased to 3200 (which is max) but the problem was the same, only took a bit longer to hit it.

**Comment by hasse69** [comment]
@billziss-gh 

Yes, you were absolutely right. After having around 83 dangling file descriptors as a result of a recursive find and `ls -l`, emptying the stand-by list using RamMap got rid of them all. The delayed release() calls were received in rar2fs. But this is a problem them? What is the limit? Is there a potential risk here that the delayed close calls results in over-allocation as seen by Cygwin? And why is dokan-fuse not showing this behaviour? Is there a way to flush the stand-by list somehow without having to use the RamMap tool? I cannot expect users of rar2fs to rely on that tool for stable operation :(

**Comment by hasse69** [comment]
@janosek 

I manged to "over-allocate" number of maximum open files in Cygwin but still do not get any errors.
```
$ ls -l /proc/13128/fd | wc -l
527
```
This was using the default number of maximum open files at 256.

I am not really sure what we can try next. Possibly something in Serviio is causing massive open requests in parallel which cause problems for the Windows Standby list? What you can try is to reproduce the problem and the check /proc/<pid>/fd. It should be close to 256 if that is your current limit.
Download/run the RamMap tool (simply google it) and  empty the Standby list. Does all the files (except maybe 3 or so) in /proc/<pid>/fd disappear? 

Another thing you can try is of course to install Dokan-fuse and see if you can reproduce the issue. But before doing that please pull lastest from master branch since I have just pushed a patch there working around the pread(2) issue in Cygwin.

**Comment by janosek** [comment]
OK.. wiped my rar2fs folder, did a git clone depth = 1, got unrar 5.5.8 and recompiled everything.

I am running rammap.

I just wiped the serviio index of the rar mount, so it is in the process of removing everything.
Now I remounted the rar drive and I am in the process of indexing it and monitoring the drive.

My ulimit for open files is back to 256.

I waited until rar2fs scanned the file system and stabilize the file descripters (4 open).  

I let the open file descripters get up to 270 and then emptied the standyby list in RAMMAP and it immediately dropped to the single digits and then started climbing again.  I let it get back up to 60 and emptied the standby list again and it dropped back to single digits again.

**Comment by hasse69** [comment]
But you did not see any *Too many files open* errors?

**Comment by janosek** [comment]
OK.. first to address comments above.  I did not see the error "too many files open" before it died.  For some reason it does not always print that message when it dies.

I think I was always able to breach the 256 ulimit previously but at some unknown number of open files the process would just silently quit.  The log does not always show anything.  Sometimes it would post an error  like too many open files or forking error but not always.

The log only shows the "Stack trace" and "End stack trace" area.  No useful messaging as in your post above.

Anyway, I was mid post when I stepped away and forgot last night..

Just in case using depth=1 was incorrect,  This time I just did a regular git clone.

After rar2fs scanned through the file system, I ran a force refresh and it started indexing and the open file descriptors started increasing quickly. I see no difference in behaviour.  Still increasing.

I found a command line program that will clear the standby list
https://wj32.org/wp/software/empty-standby-list/

I created a batch file that will run every 5 minutes:
***************************
:loop
EmptyStandbyList.exe standbylist
timeout /t 300 /nobreak
goto :loop
**********************************
I cleaned out Serviio and restarted the indexing.  Every 300 seconds, the standbylist is cleared and the file descriptors drop to single digits. I let the process run over night, with it clearing the standby list every five minutes and the process is still alive!  I am clearing the Serviio index again will run it again on this veteran process.

I will need to be talked through how to compile with Dokan to test that out.  The only way I got winfsp/rar2fs to work was parsing through your conversation with billziss-gh on the issue 75 thread.

**Comment by hasse69** [comment]
Will be interesrting to see if periodcially emptying the Standby list avoids the crash, Keep us posted!
On my side I installed Serviio and pointed it to some old share I had, I must say that Serviio knocks pretty hard on the mount point. As soon as one run of indexing was complete a new one started not too long after. 

But, something I did to make the Standby list a bit less occupied was this
```
./rar2fs --seek-length=1 -d -f /cygdrive/v/export/ -o uid=-1,gid=-1 -o 'VolumePrefix=\\test\\j' J:
```
Note the option `--seek-length=1`.

Please try that too without running your peridic script. 
If this helps also in your case we know a lot more about the problem.

**Comment by janosek** [comment]
To purge, delete the rar mount from the library/shared folders.  You will see the number under the videos icon (assuming that's what you indexed)  in the "Welcome to Serviio"  page decrease until it stabilizes.  Then you just re-add the mount.

So I got home and rar2fs was still active with my batch file clearing the standby list every 5 minutes!  

Now I am running rar2fs from the console without the batch file running and with --seek-length=1.

I will clear the mount point again and re-index and see what happens.

**Comment by janosek** [comment]
Initial observation with --seek-length=1 is that there is no difference. Within a minute I am at 77 open file descriptors.

156 after 2 ish minutes.

**Comment by hasse69** [comment]
That is expected. The Standby list is still there.
My mount has been up for over 24h now and the number of open files is stable around 600.

**Comment by janosek** [comment]
I am now at 752 and climbing,  but I have more files than the max ulimit of 3200.

**Comment by janosek** [comment]
Also note that I am set to 256
$ ulimit -a
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
open files                      (-n) 256
pipe size            (512 bytes, -p) 8
stack size              (kbytes, -s) 2032
cpu time               (seconds, -t) unlimited
max user processes              (-u) 256
virtual memory          (kbytes, -v) unlimited

**Comment by hasse69** [comment]
I would expect your number of open files to be greater than what I see because I have a hunch your library is bigger than mine. Note that with a seek length of 1 your indexing will be multitudes faster so you should see those numbers increase rapidly. 
How the limit in Cygwin works beats me, seem: almost completely bogus or works together with the Standby list which in reality is not open files.

**Comment by hasse69** [comment]
A bit off  topic perhaps, but maybe related. Can you please test my new patch also that removes the forking in rar2fs? I do not think it has anything to do with this issue but getting rid of the forking is not a bad thing and I need to stress it. Your use-case is perfect for the job :)

[patch2.txt](https://github.com/hasse69/rar2fs/files/1259268/patch2.txt)

**Comment by hasse69** [comment]
For whatever reason I cannot seem to reproduce the issue I saw before. I am starting to think I ran out of RAM and that was why it behaved like it did. After a reboot of my Windows machine and starting with a fresh mount point it has been re-indexing numerous times now for the past 24h without any issues what so ever. I think I need to rely on your setup because mine is most likely not enough to trigger this problem easily. That said, the issue as such has nothing to do with rar2fs as I see it. To me everything points to the Standby list since you have successfully been running with a periodic flush of it. Lets collect what you have after running with seek length 1 for a while. I think there are some conclusions we can draw now regarding this issue.

**Comment by janosek** [comment]
OK.. much to my surprise, with -seek-length=1, the process is still alive in the morning after a full index!
The open files are  steady at around ~1350.

I will drop the share and then reindex to see if it still exists later on.

**Comment by janosek** [comment]
That patch does not work... I run the following command and in less than 10 seconds I get "Aborted"
$ rar2fs --seek-length=1 -d -f /cygdrive/d/files/seed -o uid=-1,gid=-1 -o 'VolumePrefix=\\Janosek-pc\\x' X:
rar2fs: PID 97652: mounted X:
The service rar2fs has been started.
Aborted

**Comment by hasse69** [comment]
> That patch does not work... 
Got it. Thanks for testing. Just restore back to the unpatched version. The patch itself has nothing to do with this issue. Simply something I needed to try out on a large setup.

**Comment by janosek** [comment]
OK.  Reverted back to unpatched and running with the --seek-length=1.    I will run that for a bit unless you need me to test another patch, which I am happy to do.

**Comment by hasse69** [comment]
Since you are so eager to test stuff, here is another one for you :)
I actually got the same error as you on Linux, simply was not 100% reproducible so it slipped through. I would never send you crap like that ;)
But, please stick to what you are doing right now with testing the stability during indexing in Serviio.
I have already made some obvious conclusions based on all your efforts to test this, but I will hold those until you feel you cannot do more with what you have got.

[patch3.txt](https://github.com/hasse69/rar2fs/files/1261778/patch3.txt)

**Comment by janosek** [comment]
OK.. the process failed this latest time using --seek-length=1.  It was gone when I got home, so so far the most reliable thing that has been done is clearing the standby list every 5 minutes..

I will test patch3 now.

**Comment by janosek** [comment]
reindexing with patch3 and --seek-length=1.  Indexing starts a lot sooner after I start the rar2fs process, which is nice.  I think I will keep that option for that reason alone.

Otherwise, looks like the same behaviour.  I will let you know if I get any strange errors in the log.

**Comment by hasse69** [comment]
> It was gone when I got home, so so far the most reliable thing that has been done is clearing the standby list every 5 minutes..

Yes. This is also what I expected would happen. The seek length setting basically can only delay the problem, not fix it.

So first of all, @janosek  thanks for your patience and all your efforts in testing different things to assist in the trouble-shooting. 

What rar2fs is doing when a file is opened varies a lot on what kind of archive it is. For non-compressed archives it is not so much about memory resources but rather file descriptors. For compressed archives it is that plus a lot of other stuff on top of it. More files descriptors in the form of pipe(s) and RAM for the I/O buffer. Not even mentioning what different resources are taken by libunrar itself when a file is to be extracted. All these things together rely on on simple rule, open seize resources, close/release will free them. rar2fs cannot free resources without receiving a release call, simply due to the fact that resources are *owned* by whatever opened the file. So, the simple truth here is, no release calls, resources remain allocated. They are not leaking, because the definition of a leak is a resource that can never be freed because reference to it is lost. But they are wasted and blocked if whatever opened the file chose not to release it for a very long time. This is where this *Standby* list starts to come into play. Since it seems that the Standby list is actually holding resources without actually knowing what and how they are used for some unknown time, even if the actual owner of the resource actually did everything it could to release it, is what I would say the real culprit here. That and only that. The algorithm applied by this Standby list seems weak to say the least. If there is an enormous amount of file descriptors being allocated, should not a sane cache start to at least increase the flush-rate until things calm down? That is what I would have expected at least. The simple fact that having a script that periodically flush the Standby list seems to, in some extent, workaround the issue is evidence enough for me at least to say that this is not a problem related to rar2fs. It would hit any application that rely on open/close pairs.
So what are the options here, well I see only two right now. One is to rely on the workaround and the script that periodically flush the Standby list. The other being to investigate if there is any possibility that the WinFSP implementation of the fuse layer could ignore the Standby list and call release internally. 
But the latter is really a question to @billziss-gh. Since dokan-fuse seems to have an implementation that strictly calls release whenever an application does close on the file descriptor and seems to ignore the Standby list (and please correct me if I am wrong here) it should be feasible to expect the same behavior in WinFSP, or? I am not saying there is a bug in WinFSP, but falling back on the Standby list seems to have really nasty side-effects? 
The obvious drawback with the script as a workaround is that it will flush the list system wide. So whatever purpose the Standby list actually does have in Windows, you will effectively destroy it!

From my side I would say this issue can be closed. There is nothing that can (or even should) be done in rar2fs  in an attempt to workaround the delayed release calls. It would not solve anything, at best it can only prolong what is soon to happen anyway.

**Comment by hasse69** [comment]
@janosek 
>  Indexing starts a lot sooner after I start the rar2fs process, which is nice. I think I will keep that option for that reason alone.

I am not too sure about that being the effect of that option. The purpose of the option is that if you have a RAR volume of, lets say 100 files, only the header in the first few files will be searched for file meta data. That speeds up the listing of RAR archives enormously. But rar2fs is also caching data so the speed benefit is basically only valid at first access.

**Comment by janosek** [comment]
Should we at least keep this issue open until @billziss-gh gets back from vacation to see if he would be willing to create a standby list by-pass "feature" for us?

**Comment by janosek** [comment]
Here is the result of patch3:

The service rar2fs has been started.
rar2fs: PID 210416: mounted X:
The service rar2fs has been started.
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Bad address
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Resource temporarily unavailable
extract_callback: write: Broken pipe
extract_callback: write: Broken pipe
extract_callback: write: Bad address
extract_callback: write: Broken pipe
      0 [unknown (0x3498C)] rar2fs 210416 C:\cygwin\usr\local\bin\rar2fs.exe: *** fatal error - Internal error: TP_NUM_C_BUFS too small: 50
/home/janosek/rar2fs_start.sh: line 3: 210416 Hangup                  rar2fs --seek-length=1 -d -f /cygdrive/d/files/seed -o uid=-1,gid=-1 -o 'VolumePrefix=\\Janosek-pc\\x' X:

**Comment by hasse69** [comment]
> Here is the result of patch3:
Thanks. Will check it out. Never got those myself after many tries.

**Comment by hasse69** [comment]
@janosek 

> Should we at least keep this issue open until @billziss-gh

Lets do that

**Comment by janosek** [comment]
This is interesting.  I am trying out Dokan now and I have over 600 files open while indexing. I also had Serviio silently crash on me several times too (not your problem, obviously).   So far rar2fs has not crashed, but I had an instance where I tried to play a video in a rar2fs folder and got "bad file descriptor" error in VLC.  I stopped the process and restarted and it worked, so I could not reproduce it.   

I am running rar2fs with -d option and Dokan seems to verbose, so hopefully I can catch something useful if anything bad happens in rar2fs.

**Comment by janosek** [comment]
I will get back to you on that.  Right now I want to see if I can get through a full index of the files.  Dokan seems to work differently from WinFSP.  "X:" shows up in Serviio with Dokan, but I needed to access it with UNC with WinFSP.    Also, I want to know if the crashes are consistent in Serviio.  Its happened 3 times already today.  Feels like wack-a-mole.  Serviio is stable with winfsp, rar2fs is not. rar2fs is stable with Dokan, Serviio is not.

**Comment by janosek** [comment]
Something is going off the rails with Dokan. 
The good.  It does not quietly crash
The bad: I see  "No Memory" entries  with the -d option, After the no memory message, it starts at the first folder again.

FileMatch? : [FILENAME] (null,0,308)
  =>return
FileMatch? : [FILENAME] (null,0,309)
  =>return
FileMatch? : [FILENAME] (null,0,310)
  =>return
FileMatch? : [FILENAME] (null,0,311)
  =>return
FileMatch? : [FILENAME] (null,0,312)
  =>return
FileMatch? : [FILENAME] (null,0,313)
  =>return
FileMatch? : [LESS THAN HALF WAY DOWN FOLDER LIST] (null,0,314)
  **no memory**
index to 314
index from 314
FileMatch? : . (null,314,0)
FileMatch? : .. (null,314,1)
FileMatch? : [FIRST FOLDER] (null,314,2)
FileMatch? : [FILENAME] (null,314,3)
FileMatch? : [FILENAME] (null,314,4)
FileMatch? : [FILENAME] (null,314,5)
FileMatch? : [FILENAME] (null,314,6)
FileMatch? : [FILENAME] (null,314,7)
FileMatch? : [FILENAME] (null,314,8)
FileMatch? : [FILENAME] (null,314,9)
FileMatch? : [FILENAME] (null,314,10)
FileMatch? : [FILENAME] (null,314,11)
_<SNIP>_
FileMatch? : [FILENAME] (null,314,313)
FileMatch? : [FILENAME] (null,314,314)
  =>return
FileMatch? : [FILENAME] (null,314,315)
  =>return
FileMatch? : [FILENAME] (null,314,316)
  =>return
_<SNIP>_
FileMatch? : [3/4 DOWN FOLDER LIST] (null,314,633)
  **no memory**
index to 633
index from 633
FileMatch? : . (null,633,0)
FileMatch? : .. (null,633,1)
FileMatch? : [FIRST FOLDER] (null,633,2)
<SNIP>
FileMatch? : [FILENAME] (null,633,3)
FileMatch? : [FILENAME] (null,633,4)
FileMatch? : [FILENAME] (null,633,5)
FileMatch? : [FILENAME] (null,633,6)
FileMatch? : [FILENAME] (null,633,7)
_<SNIP>_
FileMatch? : [FILENAME] (null,633,630)
FileMatch? : [FILENAME] (null,633,631)
FileMatch? : [FILENAME] (null,633,632)
FileMatch? : [FILENAME] (null,633,633)
  =>return
FileMatch? : [FILENAME] (null,633,634)
  =>return
FileMatch? : [FILENAME] (null,633,635)
  =>return
FileMatch? : [FILENAME] (null,633,636)
  =>return
FileMatch? : [FILENAME] (null,633,637)
  =>return
_<SNIP>_
FileMatch? : [FILENAME] (null,633,789)
  =>return
FileMatch? : [FILENAME] (null,633,790)
  =>return
FileMatch? : [LAST FOLDER] (null,633,791)
  =>return
index to 792
index from 792
FileMatch? : . (null,792,0)
FileMatch? : .. (null,792,1)
FileMatch? : [FIRST FOLDER] (null,792,2)
FileMatch? : [FILENAME] (null,792,3)
FileMatch? : [FILENAME] (null,792,4)
FileMatch? : [FILENAME] (null,792,5)
FileMatch? : [FILENAME] (null,792,6)
FileMatch? : [FILENAME] (null,792,7)
_<SNIP>_
FileMatch? : [FILENAME] (null,792,791)
  STATUS_NO_MORE_FILES
###Cleanup 0013
Cleanup: \

###Close 0013
Close: \

**Comment by hasse69** [comment]
And what happens if you let your batch file run in the background?

**Comment by janosek** [comment]
Interesting.  I think maybe I found a bug in Dokany.  It only sends a "global release" for X:  when -d -f commands are issued.  When I exited out, I just closed the terminal window.  I guess Dokany does not release the drive letter when you do that (WinFSP does.. I like that).  I guess I should open a ticket on their git page?

I tried to execute without the -d -f and it was just doing nothing.  I added it back and it sent a global release and then remounted.

****************************************************************
Software Install@Janosek-PC /home/janosek
$ rar2fs.exe --seek-length=1 /cygdrive/d/files/seed X: -o umask=777 -o volname=RAR2FS_DRIVE -o uncname=\\Janosek-PC\x


Software Install@Janosek-PC /home/janosek
$
****************************************************************

Software Install@Janosek-PC /home/janosek
$ rar2fs.exe --seek-length=1 -d -f /cygdrive/d/files/seed X: -o umask=777 -o volname=RAR2FS_DRIVE -o uncname=\\Janosek-PC\x
rar2fs: PID 5884: mounted X:
Dokan: debug mode on
Dokan: use stderr
AllocationUnitSize: 512 SectorSize: 512
device opened
CheckDriveLetterAvailability failed, X: is already used
Dokan Error: CheckDriveLetterAvailability Failed
send global release for X:
rar2fs: PID 5884: unmounted X:

**Comment by janosek** [comment]
Actually, it still fails silently without -d -f even after unmounting gracefully.

**Comment by hasse69** [comment]
Sorry, I am a bit off here...this is something you believe is related to rar2fs?
Anyway, any chance you were able to index using Dokan and your Standby list flush script?
Still interesting to understand if the crashes in Serviio in combination with Dokan is related to the delayed release calls or something completely different.

**Comment by janosek** [comment]
Hello,

First of all, I apologize.  This is getting messy.  We should set up two threads one for Windows 10/Cygwin 64 bit/WinFSP and another for Windows 10/Cygwin 64 bit/Dokany.  That way we won't confuse the two, since I am playing with Dokany until @billziss-gh  lets us know either way whether he can help with WinFSP. 
I think Dokany has multiple and very different issues from WinFSP.  Serviio seems to choke on it and crash without any useful log information.  There is an open ticket in the Dokany Git forums about file open dialogs crashing in unrelated programs (Exact Audio Copy).  I noticed Serviio was chugging with a file open dialog when a rar2fs mount was active.
Dokany does not crash like WinFSP because of the standby list but it does give "no memory" errors that cause strange behaviour and then Serviio crashes (but rar2fs mount still exists, but I believe it is broken because I have gotten errors on playback using VLC).  

To be honest, I am ready to give up on Dokany (unless you want me to test something).  There are too many different issues.  WinFSP is very close to working in my use case and I can seem to keep it working with the standby list clearing script.  WinFSP is winning in my books.  I really hope @billziss-gh is open to creating a work around for us.

**Comment by hasse69** [comment]
Ok, but let's not add a new issue thread related to Dokan here. Unless there are indications pointing to a problem in rar2fs.

Regarding the Standby list. Are you saying that Dokan behaves differently here than winfsp? Does emptying the Standby list using Dokan not helping in any way regarding the number of open files reported in /proc/<pid>/fd ?

**Comment by janosek** [comment]
I am testing rar2fs/Dokany with my standby list clearing script now.  I cleared the Serviio Index, started the clearing script, started rar2fs/Dokany and started a new reindexing on X:.  Lets see what happens.  Without the clearing script, Serviio crashed but rar2fs/Dokany process was still active (albeit broken with the no memory error).  The number of files open was 3 when I got back to my computer, so I think the files closed when Serviio crashed.  So far I haven't been able to get through a full indexing of the drive without Serviio crashing, so lets see if the standby list clearing script helps.  If it does not then rar2fs/Dokany is far more unstable than rar2fs/WinFSP, in my usage.  I will get back to you with results when I have some.

**Comment by janosek** [comment]
Results: Serviio crashed at around 1500 indexed files.  rar2fs/Dokany -d log shows "no memory" errors.  Standby list clearing script did not change the behaviour.  I think I will abandon Dokany now, unless you want me to try something else.  This is not an rar2fs problem and WinFSP at least works with the standby list clearing script.

**Comment by hasse69** [comment]
I would still like to know if emptying the Standby list has any visual effect on the number of open files (/proc/<pid>/fd) when using Dokany. Because my initial tests showed that it should not. If that is the case then there is hope, especially for the WinFSP solution since it is then obvious it is possible to ignore it somehow,

**Comment by hasse69** [comment]
I am starting to distrust this whole thing completely to be honest...
This is a trace I am capturing from a very simple (one folder) mount.

```
(26328) OPEN  ...
PIPE created  op=0x201109e8  r=12, w=13   0x612b45a0
(26328) OPEN  ...
PIPE created  op=0x20111070  r=12, w=15   0x612b45a0
```

What you see should be impossible, but it is definitely going to create some really weird behavior!
Two open calls, without close in between, and yet the pipe created obtains the same read file descriptor!
In this case since 12 and 13 was already opened, it should have gotten r=14, w=15. The thread id is the same so this should not be possible, but I need to dig deeper into this.

**Comment by janosek** [comment]
OK. This time when Serviio Crashed, 8 or 9 files were open.  I executed a clearing of the standby list and those files disappeared.   It seems to be closing files now but it does not work long enough  because Serviio crashes every time I try to index before it completes.  So far I have not replicated the 600 open file situation I had yesterday.  While the indexing it working it seems to release the files.  I am struggling to give you reproducible results.

**Comment by billziss-gh** [comment]
I am finally back from vacation and slowly getting through this thread.

@hasse69 
>  Yes, you were absolutely right. After having around 83 dangling file descriptors as a result of a recursive find and ls -l, emptying the stand-by list using RamMap got rid of them all. The delayed release() calls were received in rar2fs. But this is a problem then!? What is the limit? Is there a potential risk here that the delayed close calls results in over-allocation as seen by Cygwin? And why is dokan-fuse not showing this behaviour? Is there a way to flush the stand-by list somehow without having to use the RamMap tool? I cannot expect users of rar2fs to rely on that tool for stable operation :(

The OS keeps files on the standby list for caching purposes. For example, if a closed file gets reopened, file data may be satisfied from cache. The system makes its own determination when to release the files and the file system has to (unfortunately) deal with a situation of having thousands of files open.

Regardless of this I would not expect Cygwin to crash when reaching a file limit, but rather to return something like `EMFILE`.

I do not want to speculate of why dokan-fuse does not exhibit this behavior. I note that having files on the standby list is a sign of a healthy OS and file system driver.

> Do you know what VirtualProtect failed means? Is it running out of virtual memory for a given process?

[VirtualProtect](https://msdn.microsoft.com/en-us/library/windows/desktop/aa366898(v=vs.85).aspx) is a Windows API for changing page protection. Think [mprotect](https://linux.die.net/man/2/mprotect). Error code 487 is `ERROR_INVALID_ADDRESS`: "Attempt to access invalid address."

----

Some notes regarding forking: forking in Cygwin is very fragile and does not correctly clone native DLL's (such as the WinFsp one) or the native heap. I have a hack to handle [`fuse_daemonize`](https://github.com/billziss-gh/winfsp/blob/master/opt/cygfuse/cygfuse.c#L55-L73), but forking after the file system is started is guaranteed to have problems. For example the WinFsp DLL maintains a "file system" object in native heap memory for every FUSE instance and Cygwin does not know how to clone the native heap (I think).

----

As this is a long thread and I have seen my handle (@billziss-gh) mentioned a few times please ping me again if you believe there is something I can address.

**Comment by hasse69** [comment]
@billziss-gh 
Welcome back :)

> The OS keeps files on the standby list for caching purposes. For example, if a closed file gets reopened, file data may be satisfied from cache. The system makes its own determination when to release the files and the file system has to (unfortunately) deal with a situation of having thousands of files open.

Well this is were you lost me! I do not understand how this can become a file system issue?
If I do not the get the release calls how do the OS expect me to release the resources I might have allocated while files were open? How could it possibly make a judgement on how this might affect the system as a whole? If a closed file get reopened? Well then I need to allocate new resources and If there were no release call in between the resources are lost. It is a fact I cannot possible control.

**Comment by billziss-gh** [comment]
[Please note that my explanations below are simplified and my recollection of some of these OS details is somewhat rusty.]

The primary caching mechanism in the Windows OS is the "Cache Manager". When a file gets opened the FSD (File System Driver) may choose to enlist the help of the Cache Manager to speed up I/O to a file. The Cache Manager does this by caching file data in system memory and uses its own heuristics (based on system usage and memory pressure) to determine how much data to keep cached.

If all file handles (descriptors) to a file get closed, the FSD removes its in memory file "node" and consequently the Cache Manager purges all cached memory for the file node. The problem with this is that if the file gets reopened soon after, the cache will now have to be "reprimed". It is of course a very common pattern to have a single process that continuously opens and closes a file.

For this reason, the OS has the file standby list where it keeps files that have been recently closed. If one of these files gets reopened, the cache mapping is still intact (because the file was not really closed at the FSD level) and the cache does not need to be reprimed.

The standby list is by default ON (if the FSD correctly implements caching and memory mapping). The FSD may ask (nicely) for the OS to release the file, but this is not always guaranteed. WinFsp does not (in general) ask, because the standby list is usually beneficial to a file system.

Unfortunately it can create problems for file systems like rar2fs which may have a limited amount of some resource. I see the following options:

- I would like to better understand the scenario of how rar2fs runs out of files. It would be very helpful for me if you could reiterate your most recent understanding of it so I could catch up. It is particularly curious that Cygwin crashes rather than return `EMFILE`, I am wondering whether this hides a deeper problem rather than a file limit issue.

- I would consider adding an API or configuration option to WinFsp to have it always ask the OS to **not** put a file in the standby list.

**Comment by hasse69** [comment]
@billziss-gh 

Your help is really appreciated! I understand the semantics but also realize that for a file system like rar2fs this might very well cause problems. What I do not understand is why the cache is not more clever than it is and base its flushing algo on some open file rate. But never mind. I will get back to you about the resources allocated by rar2fs and how it works.

@janosek 

Please try this patch on top of master/HEAD.
Lets see if it makes any difference at all, especially for Dokan.
[patch4.txt](https://github.com/hasse69/rar2fs/files/1271654/patch4.txt)


.

**Comment by hasse69** [comment]
@billziss-gh 

To list an archive rar2fs queries libunrar for the data structures that describes it. For a RAR *volume* that could mean hundreds of files that needs to be opened/closed in order to scan each header for meta data. The RAR format does not keep information about all files in the beginning. It might be scattered around multiple files. Adding to this is the complexity of compressed archives. To extract data from such archives an I/O buffer is allocated by rar2fs to handle the data stream. The RAR algorithm cannot jump back and forth, it is made for sequential extraction based on blocks. It is not easy (or rather impossible) to jump from one file to another. The RAR extraction algorithm for compressed (and encrypted) data is not made for random access. It is supposed to be extracted start to finish as a stream and the algorithm itself is optimized for such a case only. The I/O buffer is by default 4MiB (can be set as low as 1MiB) and thus each missing release call will result in a rather huge penalty resource wise. For a few number of archives that is not a problem. But in this case we are talking about a lot of archives being accessed and partly extracted for the purpose of indexing and creating a library/database. The number of archives is unknown but it is a lot, trust me.

**Comment by billziss-gh** [comment]
@hasse69 and @janosek I am trying to piece together a failure scenario in my mind. Please correct what I am getting wrong:

- There is a directory of too many (thousands?) of RAR files.

- Rar2fs creates a *union* file system by merging the directory hierarchies found in those RAR files.

- Under some circumstances (do we know which?) rar2fs may end up opening too many RAR files and running out of file descriptors (according to Cygwin, Windows does not actually have a defined limit).

- For some reason this kills the process rather than responding with `EMFILE` during `open`.

What does "serviio" do? What role does it play in this scenario? Does it somehow scan the rar2fs file system?

**Comment by billziss-gh** [comment]
There is also something else that does not make sense as I pick information from the thread above.

When Windows opens a file in the rar2fs file system, this is a file that rar2fs conjures from data in a RAR archive. Presumably multiple rar2fs files all have their data located in a single RAR file, which needs only one file descriptor to be open. So how can you run out of Cygwin file descriptors (unless you are trying to have a RAR fd open per Windows file or you have more than 3200 RAR files open)?

Sorry for the dumb questions.

**Comment by hasse69** [comment]
@billziss-gh 

> - There is a directory of too many (thousands?) of RAR files.

Actually, this we cannot say for sure. Cygwin's open file descriptor limit seems rather mysterious. Whatever is set in `ulimit` does not really seem to echo back on the actual limit we sometimes seem to reach. In fact, I have only seen one single report of a `Too many files open` error from @janosek. So I am more leaning towards that just being one manifestation of a deeper problem. Probably some other resource like memory is the real culprit here.

> - Rar2fs creates a *union* file system by merging the directory hierarchies found in those RAR files.

Correct.

> - Under some circumstances (do we know which? is it during listing a directory?) rar2fs may end up opening too many RAR files and running out of file descriptors (according to Cygwin, Windows does not actually have a defined limit).

As said above, I do not really think the limit of number of open file descriptors itself is causing the problem. It is rather the lack of release calls to rar2fs that is having bad side-effects. But yes, the scenario in which it happens is clear, I described it briefly above and is about Serviio scanning through the entire file system to create an index data base of everything in it.

> What does "serviio" do? What role does it play in this scenario? Does it somehow scan the rar2fs file system?

Yupp! :)

> When Windows opens a file in the rar2fs file system, this is a file that rar2fs conjures from data in a RAR archive. Presumably multiple rar2fs files all have their data located in a single RAR file, which needs only one file descriptor to be open.

When Windows opens a files *it* sees, rar2fs gets an open call for that file, yes. But the file is virtual, thus it will not request to open the RAR file which holds it. Obviously that will have to be opened too but that is done internally within rar2fs. If an archive contains 100 files, from Windows there will be 100 open requests and rar2fs unfortunately needs to open its container as many times since, again, we are talking sessions of streaming data here and only opening the file ones would make each session lose context of file position etc. But to be honest, we are talking archives with very few files in it, in most cases only one so that is hardly the problem here.

**Comment by billziss-gh** [comment]
@hasse69 thanks for the explanations so far.

Can you also explain under which circumstances rar2fs will fork (after creating the FUSE file system) and what does the forked rar2fs processes do?

Forks in Cygwin will not reload native DLL's that have been dynamically loaded (e.g. the WinFsp DLL) and will not clone the default process heap where the user portion of WinFsp (i.e. the WinFsp DLL) stores data. So a forked rar2fs process is actually severely crippled.

For example, if some thread in the forked process tries to do something with FUSE, then a **BOOM** is likely.

Here is thread discussing some of the Cygwin fork limitations: [link](http://cygwin.1069669.n5.nabble.com/Fork-and-Windows-Heap-td127686.html)

**Comment by hasse69** [comment]
@billziss-gh 

> Can you also explain under which circumstances rar2fs will fork (after creating the FUSE file system) and what does the forked rar2fs processes do?

It forks only for compressed archives and in this use case they are few in comparison to the archives in pure store/uncompressed form. 

> Forks in Cygwin will not reload native DLL's that have been dynamically loaded (e.g. the WinFsp DLL) and will not clone the default process heap where the user portion of WinFsp (i.e. the WinFsp DLL) stores data. So a forked rar2fs process is actually severely crippled.

Yes, so I have been told (actually you did). That is specifically what the patches I am sending to @janosek is addressing. They use threads instead of forked processes.

**Comment by hasse69** [comment]
Another thing you need to keep in mind here is that for RAR volumes, even if rar2fs never keeps more than one file open at a time, extracting whatever is in it might result in hundreds of files being opened at least once. And then I have not even mentioned the rather odd behavior in Cygwin. If I do a simple hexdump of only a few bytes of a file on the rar2fs volume, Linux opens the file once and then close it. Cygwin opens the file four times for doing the same thing! That is not really helping out here...

**Comment by billziss-gh** [comment]
@hasse69 

> It forks only for compressed archives and in this use case they are few in comparison to the archives in pure store/uncompressed form.

I am assuming then that you do not think that issue happens because of a fork failure (because no forks actually happen in @janosek's scenario).

@janosek 

Looking back into the thread it appears that the original issue is that rar2fs crashes after it is run for some time. There is a way to get core dumps under Cygwin. Have you tried [dumper](https://cygwin.com/cygwin-ug-net/dumper.html)?

**Comment by hasse69** [comment]
@billziss-gh 

> I am assuming then that you do not think that issue happens because of a fork failure (because no forks actually happen in @janosek's scenario).

Something like that, yes. But also see my previous post, I edited a bit and mentioned what the forked processes/threads are doing.

**Comment by billziss-gh** [comment]
> That is specifically what the patches I am sending to @janosek is addressing. They use threads instead of forked processes. But I am not sure it is really a problem here because the forked processed (or threads) does nothing towards fuse. They are reader/writers to handle throttling while extracting data from an archive.

Understood.

**Comment by janosek** [comment]
@hasse69:
 
I redownloaded rar2fs and applied patch5.txt with Dokany.

It started and did the usual start up.  Serviio didn't get to indexing before I got the following message at the bottom of the log:
/home/janosek/rar2fs_dokan_start.sh: line 2: 13212 Aborted                 rar2fs.exe --seek-length=1 -d -f /cygdrive/d/files/seed X: -o umask=777 -o volname=RAR2FS_DRIVE -o uncname=\\Janosek-PC\x

There were several "no memory" errors flung out by Dokany.
I did not get any "extract_callback: write: Bad address" appearing in the log.

Is there anything you expected to see?

**Comment by billziss-gh** [comment]
@janosek thanks. Getting the core dump file will help us pinpoint the issue.

My suggestion is to carefully note the commit hash that you use and any patches you apply to rar2fs. [This will help us reproduce the exact same rar2fs binary for investigation.] You can then compile a debug version of rar2fs (presumably using the `-g` compiler switch, although @hasse69 knows better here) that uses winfsp-fuse and run it until it crashes. Then post the resulting core dump to this thread to analyze using gdb.

**Comment by janosek** [comment]
Tomorrow I will switch back to WinFSP, unless @hasse69 found something useful in my last email and wants me to continue.

**Comment by hasse69** [comment]
Feel free to go back to WinFSP. I have problems with replacing forked processes with threads due to libunrar exception handling. I am currently unable to make the program work as intended.

**Comment by janosek** [comment]
Back on WinFSP.  totally wiped the cygwin folder and uninstalled Dokany from Windows.

Current Head/Master of rar2fs is broken (no patches installed, just git clone):

[New Thread 0x3240]
[New Thread 0x2950]
[New Thread 0x2dd4]
[New Thread 0x2c7c]
[New Thread 0x1a2c]
[New Thread 0x18f8]
[New Thread 0x2e7c]
[New Thread 0x1f00]
[New Thread 0x286c]
[New Thread 0x33ac]
[New Thread 0x22a4]
[New Thread 0x1570]
[New Thread 0x1c38]
Segmentation fault (core dumped)

**Comment by hasse69** [comment]
Interesting since this is the version you have been using all along except for the broken patches I provided. Need to get a stack trace or else there is not much I can do I am afraid.

**Comment by janosek** [comment]
I should have mentioned that I am using dumper.exe on the process.
I also ran
 git reset --hard b5a385b
 assuming that would get me to an earlier version but got the same error.
I will try without dumper.exe

**Comment by billziss-gh** [comment]
@janosek dumper should produce a core dump (likely on the Cygwin root directory). Can you please post it here?

**Comment by janosek** [comment]
Hello,
This run is with b5a385b and running "dumper rar2fs.exe [WIN PID]"
It crashed in 20 seconds.  
[rar2fs.core..zip](https://github.com/hasse69/rar2fs/files/1272370/rar2fs.core.zip)

Since then, I uninstalled and reinstalled winfsp in windows, wiped the rar2fs source folder, git cloned the latest version of rar2fs and recompiled.  I am currently running WITHOUT dumper.exe on the process and it is working and Serviio is indexing.

I am seeing new bahaviour now.  Serviio is indexing but I don't see any files being opened under /proc/[PID]/fd. If I open a Windows Explorer window, I see two new handles for the base folder.  If I close windows explorer, they disappear.  I only have a single rar2fs process running.  Normally all hell breaks loose as soon as Serviio starts indexing.  I also cleared the standby list in case anything was cached.

**Comment by janosek** [comment]
Interesting.  Serviio thinks it has finished indexing, but it only indexed non-rared files. I see the entire structure in Windows explorer and files play fine. 

About 20 files were stuck open by the end.  clearing the standby list got rid of them.

I am going to start a new process and try again.

**Comment by janosek** [comment]
Thought it might be because I forgot to log the serviio process in under my non-admin account, but that didn't seem to work,  Seems to be somehow ignoring the files that were in rars, but I can see them in the share under Windows Explorer.

I also upgraded my Java version, which Serviio runs on a couple days ago while messing with Dokany.  Maybe its somehow a permission thing?

**Comment by janosek** [comment]
Just tried uninstalling the 32 bit Java I accidentally installed and installed the 64 bit version.  Same thing.

Somehow Serviio is now ignoring anything that was originally in rars and only indexing unrared videos in my share just like if I mounted the regular folder.

But if I go to X: and play a rared file in VLC, it works fine.  

Lets see what Plex does.

**Comment by janosek** [comment]
Plex Works!  Indexing rar files.  Something must have broke in Serviio with the latest Java version.  I might raise it up in their forums.

**Comment by janosek** [comment]
So far rar2fs/WinFSP/Plex is not going off the rails.  I see file opening and closing and not accumulating.

**Comment by hasse69** [comment]
Ok, a bit frustrating with this extremely non-deterministic behavior you have experienced recently.

**Comment by janosek** [comment]
Frustrating indeed.  I cannot get Serviio to see the videos in processed rar files no matter what version I go to.  I uninstalled all versions of java.  installed old ones, etc.
I even tried removing the --seek-length=1 (which causes problems with VLC and subtitles).  

Plex is working like a charm, so maybe I will switch.

I don't know what to say?  Close the thread and maybe it will come back?  If it continues to work like this, I will be happy!

**Comment by hasse69** [comment]
@janosek 
> I even tried removing the --seek-length=1 (which causes problems with VLC and subtitles).

Use a value of 2 instead. Or use the per archive configuration option. See `rarconfig.example` for details.

> I don't know what to say? Close the thread and maybe it will come back? If it continues to work like this, I will be happy!

Lets keeps it open for a while. Does no harm.

**Comment by janosek** [comment]
So Java suddenly does not like the rar2fs file system.  I tried reverting to the pre-pread fix, I tried uninstalling and going back to an old version of serviio.  VERY frustrating.  It simply won't index videos that originate from rar files now.  It sees the file system.  If I tell it to index picture, it will index pictures in the same folder as the videos from rar files.  

I installed mucommander, which is a java based file manager to see if it has strange behaviour, and it does.  When I double click on a video from a rar file on my X drive, it does nothing.  When I double click on an nfo, it opens.  When I double click on a sample video not from a rar file, it opens.  when I select open natively, it does nothing.  When I select "Reveal in Explorer" and double click on a rar-based video, it plays just fine in VLC.   Windows is a very frustrating operating system.

**Comment by hasse69** [comment]
@janosek 
Still sounds like some access right issue? I guess explorer is executed as another account thanyour local user? Also, it would be interesting to see if you get any calls what so ever into rar2fs when accessing files through Java.  You can enable debug level 3 (`./configure --enable-debug=3 ...`) and rebuild. Running using `-f` should not drop a few lines on the console.
When it comes to permissions etc and WinFSP @billziss-gh is really the guy to answer.

**Comment by janosek** [comment]
Back in business! The problem was a remnant of Dokan, I believe.  The base folder was being shared as X drive and that is what serviio was seeing, but not the rar2fs mount, hence why no videos in rar files were being indexed.  I finally discovered it when I decided to mount a SINGLE rar folder rather than the whole library and Serviio was still indexing files in the entire library.  Now it is indexing the rar based videos. and the file descriptor count is increasing quickly.

I recompiled with --enable-debug=3 as well.

Everytime I try to run dumper rar2fs.exe [WIN PID] it crashes rar2fs within a minute or so.  Any ideas on this?

I was fiddling with the Winfsp command line options that rar2fs passes.  I currently have the following:
rar2fs -d -f --seek-length=2 /cygdrive/d/files/seed Y: -oVolumePrefix='\\\Janosek-pc\\y',uid=-1,gid=-1,umask=000

the umask=000 seemed to change the behaviour of mucommander.  When I double clicked on a rar-based video, it started playing.  I am keeping that now.

**Comment by billziss-gh** [comment]
According to the [dumper](https://cygwin.com/cygwin-ug-net/dumper.html) page the CYGWIN environment variable needs to be:

```
error_start=x:\path\to\dumper.exe
```

So I would try something along the lines of:

```
export CYGWIN="error_start=C:\\cygwin\\bin\\dumper"

# or if CYGWIN is not empty

export CYGWIN="$CYGWIN error_start=C:\\cygwin\\bin\\dumper"
```

[Notice the double backslashes.]

Re: permissions in WinFsp-FUSE. The options `uid=-1,gid=-1` will make files owned by the user/group who launches the file system. The `umask=000` will effectively give access `0777` to everyone. So you should not have permission problems with these options.

**Comment by janosek** [comment]
So after multiple attempts to make rar2fs crash, I cannot.  That said I have new behaviour that persisted after a couple restarts.  I dumped all my videos into one folder try to and make rar2fs die.  What I found was the open files climbed to 3200 and stayed there and Serviio stopped indexing.  After a few hours, 50 or so files closed.  I hit the force refresh button and a bunch more files (50 or so I would guess) were indexed and then it stopped again and the rar2fs process again had 3200 files open.

I just cleared the standby list and it dropped down to 4 files... and now files are indexing again after I clicked force refresh again.

**Comment by hasse69** [comment]
So you are back to a solution with periodically clearing the Standby list, right?
I am stll curious though to why the Standby list seems to be by-passed when using Dokany.
@billziss-gh Can there be such a major differences here between the two?

**Comment by billziss-gh** [comment]
> Can there be such a major differences here between the two?

The call to "ask" the OS to really release a file is  [`MmForceSectionClosed`](https://msdn.microsoft.com/en-us/library/windows/hardware/ff549811(v=vs.85).aspx). Dokany does not appear to call it according to GitHub search. So it may be that its files are never entered in the list.

In the past Dokany had many issues with memory mapped files. I no longer follow Dokany development and am not sure if those persist or not. Dokany also does not support the NTOS Cache Manager. It may be that Windows chooses not to enter files in the standby list unless they look a certain way. After all the standby list is primarily a cache for memory mapped files and an FSD that does not implement caching may not get their benefit.

As mentioned earlier having files in the standby list is a sign of a properly functioning FSD, so I believe WinFsp is doing the right thing here.

I am open to add a method to disable the standby list for file systems like rar2fs. OTOH, I am still not convinced that we know what the true issue is.

**Comment by hasse69** [comment]
@billziss-gh

> After all the standby list is primarily a cache for memory mapped files and an FSD that does not implement caching may not get their benefit.

Interesting. Are you saying that if I do not use cygwin/mmap() in my code the Standby list should not be affected as much? Or is this purely Windows internal? 

> I am open to add a method to disable the standby list for file systems like rar2fs. OTOH, I am still not convinced that we know what the true issue is.

That sound great! Regarding the issue. If you are referring to the crash, no, we do not really know what caused it. All I know is that I have not made any patches in rar2fs to address it. When it comes to the number of files issue I am fairly sure we know by now that in the way Windows use the Standby list rar2fs (and probably also many other FUSE based file systems) is hitting a resource allocation problem due to the delayed release calls.

**Comment by billziss-gh** [comment]
> Interesting. Are you saying that if I do not use cygwin/mmap() in my code the Standby list should not be affected as much? Or is this purely Windows internal?

This is purely Windows internal. You have no control over it with Cygwin or even any Win32 API. This is something that can only be done at the FSD (File System Driver) layer.

> > I am open to add a method to disable the standby list for file systems like rar2fs.
>
>That sounds great!

Feel free to open an enhancement request in WinFsp. I cannot promise a quick fix though as I am quite busy at the moment...

**Comment by janosek** [comment]
So, @hasse69, will you open an enhancement request? I could try, but I may not be precise enough in my request.

I am still surprised that the behavior changed in that there is no crash, the number of files open filled to 3200 and then it seemed to hit a wall where rar2fs could no longer work properly (but did not crash).  Are you sure your pread(2) patch did not change anything for WinFSP?

In any case, I have a work around with my clear_standby_list.bat file:

:loop
    runas /user:software_install /savecred "D:\rar2fs_docs\EmptyStandbyList.exe standbylist"
timeout /t 300 /nobreak
goto :loop

A bit heavy handed, but it will work for now.  I have it loading minimized on boot up along with rar2fs.

**Comment by hasse69** [comment]
@janosek 

> So, @hasse69, will you open an enhancement request? I could try, but I may not be precise enough in my request.

I think you can give it a shot. After all, you have all the results from your tests and the script to workaround it. And @billziss-gh also knows about this already.

> I am still surprised that the behavior changed in that there is no crash, the number of files open filled to 3200 and then it seemed to hit a wall where rar2fs could no longer work properly (but did not crash). 

Yes I am also surprised because this is exactly how it should behave. 3200 open files seems to be some hard limit in Windows? The limit you set in Cygwin could actually be fake but I would not be able to answer that. If you hit the limit rar2fs should be robust enough to report the error to the higher layer and then the higher layer should in turn handle it gracefully. That seems to suddenly be the case?

> Are you sure your pread(2) patch did not change anything for WinFSP?

You can never be sure, but it would certainly surprise me. WinFSP does not really care about this. It is the application using the file system that would/should be affected by the patch since EOF hopefully is reported properly now despite the bug in Cygwin.

**Comment by hasse69** [comment]
I will keep this issue open a while for follow-up, but will lower the priority due to that rar2fs is most likely not in error here as I can see it.

**Comment by billziss-gh** [comment]
> Yes I am also surprised because this is exactly how it should behave. 3200 open files seems to be some hard limit in Windows? The limit you set in Cygwin could actually be fake but I would not be able to answer that. If you hit the limit rar2fs should be robust enough to report the error to the higher layer and then the higher layer should in turn handle it gracefully. That seems to suddenly be the case?

The 3200 limit almost certainly comes from Cygwin's emulation of file descriptors. Windows does not have any defined limit on how many files a process can open. There is the practical limit of running out of handles or out of system memory for the files but these are not hard-coded values.

> In any case, I have a work around with my clear_standby_list.bat file:
>
> :loop
> runas /user:software_install /savecred "D:\rar2fs_docs\EmptyStandbyList.exe standbylist"
> timeout /t 300 /nobreak
> goto :loop

BTW, there has been some anecdotal evidence that a filter driver named superfetch (or similar name in other versions of Windows) may be responsible for keeping files in the standby list for long times. You may be able to get rid of your batch file by simply disabling this service.

https://forum.sysinternals.com/rammap-empty-standby-list_topic27297_post132769.html

**Comment by janosek** [comment]
I stopped and disabled the superfetch process  in Windows 10 x64 and restarted the indexing.  rar2fs is at 3200 open files, so it didn't change the behaviour.

**Comment by hasse69** [comment]
What strikes me odd is why it stops at 3200 when your ulimit in fact is set to 256 in Cygwin?

**Comment by hasse69** [comment]
I will now close this issue. I do not think there is much more to add currently.
When/if there is a patch for WinFSP as requested in https://github.com/billziss-gh/winfsp/issues/104 and it still does not work as expected, feel free to file a new issue.
